src/toil/lib/retry.py

```diff
@@ -13,31 +13,145 @@
 # limitations under the License.
 
 # 5.14.2018: copied into Toil from https://github.com/BD2KGenomics/bd2k-python-lib
-
-from __future__ import absolute_import
-
-from future import standard_library
-standard_library.install_aliases()
-from builtins import next
-from contextlib import contextmanager
 import time
-import urllib.request, urllib.error, urllib.parse
-try:
-    from httplib import BadStatusLine
-except ImportError:
-    # This has moved in Python 3
-    from http.client import BadStatusLine
+import copy
+import functools
 import logging
 
-log = logging.getLogger( __name__ )
+from contextlib import contextmanager
+from typing import List, Set, Optional, Tuple, Callable, Any
+
+log = logging.getLogger(__name__)
+
+
+class ErrorCondition:
+    """
+    A wrapper describing an error condition.
+
+    ErrorCondition events may be used to define errors in more detail to determine whether to retry.
+
+    :param error: An Exception (required)
+    :param error_codes: Error codes that must match to be retried (optional; defaults to not checking)
+    :param error_message_must_include: A string that must be in the error message to be retried
```

Comment:
It is since that's the only case where this was used, though I've thought about using the traceback library to do a full scan.  I'll attempt to implement that instead.